app/src/ui/delete-branch/delete-branch-dialog.tsx

```diff
@@ -61,7 +61,10 @@ export class DeleteBranch extends React.Component<
   }
 
   private renderDeleteOnRemote() {
-    if (this.props.branch.remote && this.props.existsOnRemote) {
+    if (
+      this.props.branch.remote &&
+      getBranchAheadBehind(this.props.repository, this.props.branch) !== null
```

Comment:
There are a couple problems here:

1. At a high level, the app is using the [Flux](https://facebook.github.io/flux/) architecture. So all of our state should exist at the top level and then get passed down. We generally shouldn't be pulling data from git within a component itself.
1. On a more practical level, `getBranchAheadBehind` returns a `Promise`, so this is really just checking to see if the returned `Promise` isn't `null`, which it never is.

Reply:
Right, understood. I've moved the logic on if the branch exists on the remote to when the `Branch` model is instantiated in e606e8bea4de5f68d6b6c28e77ea575b81fb57f9.

My only thought is is instead of a `show-ref` git operation for each branch with an upstream we could use `ls-remote` at the start of collecting all the branches and make our own checks whether the upstream exists in that list?